core/src/reporter/state.ts

```diff
@@ -225,27 +239,93 @@ export class State {
       }
     })
 
-    const wallets = reporters.map((R) => {
-      let wallet
+    const wallets: { [key: string]: Wallet }[] = []
+    for (const R of reporters) {
+      let wallet: Wallet
+      let nonce: number
       if (this.delegatedFee) {
         wallet = buildCaverWallet({
           privateKey: R.privateKey,
           providerUrl: this.providerUrl
         })
+        nonce = Number(await wallet.caver.rpc.klay.getTransactionCount(wallet.address))
       } else {
         wallet = buildWallet({
           privateKey: R.privateKey,
           providerUrl: this.providerUrl
         })
+        nonce = await wallet.getTransactionCount()
       }
 
-      return { [R.oracleAddress]: wallet }
-    })
+      wallets.push({ [R.oracleAddress]: wallet })
+
+      // Update nonce
+      this.nonces = {
+        ...this.nonces,
+        [R.oracleAddress]: nonce
+      }
+
+      // Create a mutex object
+      this.mutex = new Mutex()
+    }
 
     // Update
     await this.redisClient.set(this.stateName, JSON.stringify(reporters))
     this.wallets = Object.assign({}, ...wallets)
 
     return reporters
   }
+
+  /**
+   * Get nonce for oracleAddress. If nonce is not found, raise an error.
+   * This function implements a mutex to ensure it cannot be called concurrently.
+   *
+   * @param {string} oracleAddress
+   * @return {number} nonce
+   * @exception {OraklErrorCode.WalletNotActive} raise when wallet is not active
+   * @exception {OraklErrorCode.FailedToGetWalletTransactionCount} raise when failed to get wallet transaction count
+   */
+  async getAndIncrementNonce(oracleAddress: string): Promise<number> {
+    return await this.mutex.runExclusive(async () => {
+      const wallet = this.wallets[oracleAddress]
+      if (!wallet) {
+        const msg = `Wallet for oracle ${oracleAddress} is not active`
+        this.logger.error({ name: 'getAndIncrementNonce', file: FILE_NAME }, msg)
+        throw new OraklError(OraklErrorCode.WalletNotActive, msg)
+      }
+
+      let remoteNonce: number
+      try {
+        if (this.delegatedFee) {
+          const caverWallet = wallet as CaverWallet
+          remoteNonce = Number(
+            await caverWallet.caver.rpc.klay.getTransactionCount(caverWallet.address)
+          )
+        } else {
+          remoteNonce = await (wallet as NonceManager).getTransactionCount()
+        }
+      } catch (error) {
+        const msg = `Failed to get nonce for wallet`
+        this.logger.error({ name: 'getAndIncrementNonce', file: FILE_NAME }, msg)
+        throw new OraklError(OraklErrorCode.FailedToGetWalletTransactionCount, msg)
+      }
+
+      const localNonce = this.nonces[oracleAddress]
+
+      let nonce: number
+      if (!localNonce || remoteNonce > localNonce) {
```

Comment:
@Intizar-T Just curious, when `remoteNonnce < localNonce` it is okay? Do we trust `localNonce` more than `remoteNonce`?